src/readonly.rs

```diff
@@ -0,0 +1,165 @@
+    writeable: bool,
+}
+
+impl<'py, T: TypeNum, D: Dimension> PyReadOnlyArray<'py, T, D> {
+    /// Returns the immutable view of the internal data of `PyArray` as slice.
+    ///
+    /// Returns `ErrorKind::NotContiguous` if the internal array is not contiguous.
+    /// # Example
+    /// ```
+    /// use numpy::{PyArray, PyArray1};
+    /// use pyo3::types::IntoPyDict;
+    /// let gil = pyo3::Python::acquire_gil();
+    /// let py = gil.python();
+    /// let py_array = PyArray::arange(py, 0, 4, 1).reshape([2, 2]).unwrap();
+    /// let readonly = py_array.readonly();
+    /// assert_eq!(readonly.as_slice().unwrap(), &[0, 1, 2, 3]);
+    /// let locals = [("np", numpy::get_array_module(py).unwrap())].into_py_dict(py);
+    /// let not_contiguous: &PyArray1<i32> = py
+    ///     .eval("np.arange(10)[::2]", Some(locals), None)
+    ///     .unwrap()
+    ///     .downcast()
+    ///     .unwrap();
+    /// assert!(not_contiguous.readonly().as_slice().is_err());
+    /// ```
+    pub fn as_slice(&self) -> Result<&[T], NotContiguousError> {
+        unsafe { self.array.as_slice() }
+    }
+
+    /// Get the immutable view of the internal data of `PyArray`, as
+    /// [`ndarray::ArrayView`](https://docs.rs/ndarray/latest/ndarray/type.ArrayView.html).
+    ///
+    /// # Example
+    /// ```
+    /// # #[macro_use] extern crate ndarray;
+    /// use numpy::PyArray;
+    /// let gil = pyo3::Python::acquire_gil();
+    /// let py = gil.python();
+    /// let array = PyArray::arange(py, 0, 4, 1).reshape([2, 2]).unwrap();
+    /// let readonly = array.readonly();
+    /// assert_eq!(readonly.as_array(), array![[0, 1], [2, 3]]);
+    /// ```
+    pub fn as_array(&self) -> ArrayView<'_, T, D> {
+        unsafe { self.array.as_array() }
+    }
+}
+
+/// one-dimensional readonly array
+pub type PyReadOnlyArray1<'py, T> = PyReadOnlyArray<'py, T, Ix1>;
+/// two-dimensional readonly array
+pub type PyReadOnlyArray2<'py, T> = PyReadOnlyArray<'py, T, Ix2>;
+/// three-dimensional readonly array
+pub type PyReadOnlyArray3<'py, T> = PyReadOnlyArray<'py, T, Ix3>;
+/// four-dimensional readonly array
+pub type PyReadOnlyArray4<'py, T> = PyReadOnlyArray<'py, T, Ix4>;
+/// five-dimensional readonly array
+pub type PyReadOnlyArray5<'py, T> = PyReadOnlyArray<'py, T, Ix5>;
+/// six-dimensional readonly array
+pub type PyReadOnlyArray6<'py, T> = PyReadOnlyArray<'py, T, Ix6>;
+/// dynamic-dimensional readonly array
+pub type PyReadOnlyArrayDyn<'py, T> = PyReadOnlyArray<'py, T, IxDyn>;
+
+impl<'py, T: TypeNum, D: Dimension> FromPyObject<'py> for PyReadOnlyArray<'py, T, D> {
+    fn extract(obj: &'py PyAny) -> PyResult<Self> {
+        let array: &PyArray<T, D> = obj.extract()?;
+        Ok(PyReadOnlyArray::from(array))
+    }
+}
+
+impl<'py, T, D> IntoPy<PyObject> for PyReadOnlyArray<'py, T, D> {
+    fn into_py(self, py: Python<'_>) -> PyObject {
+        let PyReadOnlyArray {
+            array,
+            writeable: _,
+        } = self;
```

Comment:
I think this might also work:

```suggestion
        let PyReadOnlyArray { array, .. } = self;
```